File: multimodal-generation/image-generation/javascript/amazon_nova_canvas_utils.js
const fs = require('fs').promises;
const path = require('path');
const { BedrockRuntimeClient, InvokeModelCommand } = require('@aws-sdk/client-bedrock-runtime');

/**
 * Loads an image from disk and returns a Base64 encoded string.
 * @param {string} imagePath - The path to the image file.
 * @returns {Promise<string>} A Base64 encoded string
 */
async function loadImageAsBase64(imagePath) {
    const imageBuffer = await fs.readFile(imagePath);
    return imageBuffer.toString('base64');
}

/**
 * Converts a Base64 encoded string to a Buffer.
 * @param {string} base64Str - A Base64 encoded string.
 * @returns {Buffer} A Buffer containing the image data.
 */
function base64ToBuffer(base64Str) {
    // Remove the data URL prefix if it exists (e.g., 'data:image/jpeg;base64,')
    if (base64Str.includes(',')) {
        base64Str = base64Str.split(',')[1];
    }
    
    return Buffer.from(base64Str, 'base64');
}

/**
 * Generates images using Amazon Nova Canvas model.
 * @param {Object} inferenceParams - The inference parameters for image generation.
 * @param {string} baseFilename - Base filename for saved files.
 * @param {string} saveFolderPath - Path to save generated files.
 * @param {string} modelId - The model ID to use.
 * @param {string} regionName - AWS region name.
 * @param {string} endpointUrl - Optional custom endpoint URL.
 * @returns {Promise<Object>} The response body containing generated images.
 */
async function generateImages(
    inferenceParams,
    baseFilename = '',
    saveFolderPath = null,
    modelId = 'amazon.nova-canvas-v1:0',
    regionName = 'us-east-1',
    endpointUrl = null
) {
    // If the caller has provided a save folder path, save the inference params to disk
    if (saveFolderPath) {
        await fs.mkdir(saveFolderPath, { recursive: true });
        
        // Save the inference params
        await fs.writeFile(
            path.join(saveFolderPath, `${baseFilename}inference_params.json`),
            JSON.stringify(inferenceParams, null, 2)
        );
    }

    let imageCount = 1;
    if (inferenceParams.imageGenerationConfig?.numberOfImages) {
        imageCount = inferenceParams.imageGenerationConfig.numberOfImages;
    }

    console.log(`Generating ${imageCount} image(s) with ${modelId}`);

    // Display the seed value if one is being used
    if (inferenceParams.imageGenerationConfig?.seed) {
        console.log(`Using seed: ${inferenceParams.imageGenerationConfig.seed}`);
    }

    const clientConfig = {
        region: regionName,
        requestTimeout: 300000
    };
    
    if (endpointUrl) {
        clientConfig.endpoint = endpointUrl;
    }

    const bedrock = new BedrockRuntimeClient(clientConfig);
    
    const startTime = new Date();

    try {
        
        const command = new InvokeModelCommand({
            body: JSON.stringify(inferenceParams),
            modelId: modelId,
            accept: 'application/json',
            contentType: 'application/json'
        });

        const response = await bedrock.send(command);
        const duration = new Date() - startTime;
        console.log(`Image generation took ${(duration / 1000).toFixed(2)} seconds.`);

        const responseMetadata = response.$metadata;

        // Log the request ID
        console.log(`Image generation request ID: ${responseMetadata.requestId}`);

        // Write response metadata to disk
        if (saveFolderPath) {
            await fs.writeFile(
                path.join(saveFolderPath, `${baseFilename}response_metadata.json`),
                JSON.stringify(responseMetadata, null, 2)
            );
        }

        const responseBody = JSON.parse(new TextDecoder().decode(response.body));

        // Check for non-exception errors
        if (responseBody.error && saveFolderPath) {
            await fs.writeFile(
                path.join(saveFolderPath, `${baseFilename}error.txt`),
                responseBody.error
            );
        }

        // Write the images to disk
        if (saveFolderPath && responseBody.images) {
            for (let index = 0; index < responseBody.images.length; index++) {
                const imageBase64 = responseBody.images[index];
                const imageBuffer = base64ToBuffer(imageBase64);
                await fs.writeFile(
                    path.join(saveFolderPath, `${baseFilename}image_${index}.png`),
                    imageBuffer
                );
            }
        }

        return responseBody;

    } catch (error) {
        // Write the error message to disk
        if (saveFolderPath) {
            await fs.writeFile(
                path.join(saveFolderPath, `${baseFilename}error.txt`),
                error.toString()
            );
        }
        throw error;
    }
}

// // For compatibility with the VirtualTryOn script
// function base64ToPilImage(base64Str) {
//     // JavaScript doesn't have PIL, so we'll just return the base64 string
//     // In a real implementation, you'd convert this to an image format your environment can display
//     return base64Str;
// }

module.exports = {
    loadImageAsBase64,
    base64ToBuffer,
    generateImages,
    //base64ToPilImage
};

File: multimodal-generation/image-generation/python/amazon_nova_canvas_utils.py
import base64
import io
import json
import logging
import os
from datetime import datetime, time

import boto3
from botocore.config import Config
from PIL import Image

logger = logging.getLogger(__name__)


def load_image_as_base64(image_path):
    """
    Loads an image from disk and returns a Base64 encoded string.

    Args:
        image_path (str): The path to the image file.

    Returns:
        str: A Base64 encoded string
    """
    with open(image_path, "rb") as image_file:
        return base64.b64encode(image_file.read()).decode("utf-8")


def base64_to_pil_image(base64_str):
    """
    Converts a Base64 encoded string to a PIL Image object.

    Args:
        base64_str (str): A Base64 encoded string.

    Returns:
        PIL.Image: A PIL Image object.
    """
    # Remove the data URL prefix if it exists (e.g., 'data:image/jpeg;base64,')
    if "," in base64_str:
        base64_str = base64_str.split(",")[1]

    # Decode base64 string to bytes
    image_bytes = base64.b64decode(base64_str)

    # Create a BytesIO object from the bytes
    image_buffer = io.BytesIO(image_bytes)

    # Create PIL Image object
    pil_image = Image.open(image_buffer)

    return pil_image


def generate_images(
    inference_params,
    base_filename="",
    save_folder_path=None,
    model_id="amazon.nova-canvas-v1:0",
    region_name="us-east-1",
    endpoint_url=None,
):
    # If the caller has provided a save folder path, save the inference params
    # to disk.
    if save_folder_path:
        os.makedirs(save_folder_path, exist_ok=True)

        # Save the inference params.
        with open(
            os.path.join(save_folder_path, f"{base_filename}inference_params.json"), "w"
        ) as f:
            json.dump(inference_params, f, indent=2)

    image_count = 1
    if "imageGenerationConfig" in inference_params:
        if "numberOfImages" in inference_params["imageGenerationConfig"]:
            image_count = inference_params["imageGenerationConfig"]["numberOfImages"]

    logger.info(f"Generating {image_count} image(s) with {model_id}")

    # Display the seed value if one is being used.
    if "imageGenerationConfig" in inference_params:
        if "seed" in inference_params["imageGenerationConfig"]:
            logger.info(
                f"Using seed: {inference_params['imageGenerationConfig']['seed']}"
            )

    bedrock_client_optional_args = (
        {} if endpoint_url is None else {"endpoint_url": endpoint_url}
    )
    bedrock = boto3.client(
        service_name="bedrock-runtime",
        region_name=region_name,
        config=Config(read_timeout=300),
        **bedrock_client_optional_args,
    )

    body_json = json.dumps(inference_params, indent=2)

    start_time = datetime.now()

    try:
        response = bedrock.invoke_model(
            body=body_json,
            modelId=model_id,
            accept="application/json",
            contentType="application/json",
        )
        duration = datetime.now() - start_time
        logger.info(
            f"Image generation took {round(duration.total_seconds(), 2)} seconds."
        )

        response_metadata = response.get("ResponseMetadata")

        # Log the request ID.
        logger.info(
            f"Image generation request ID: {response['ResponseMetadata']['RequestId']}"
        )

        # Write response metadata to disk.
        if save_folder_path:
            with open(
                os.path.join(
                    save_folder_path, f"{base_filename}response_metadata.json"
                ),
                "w",
            ) as f:
                json.dump(response_metadata, f, indent=2)

        response_body = json.loads(response.get("body").read())

        # Check for non-exception errors.
        if "error" in response_body and save_folder_path:
            with open(
                os.path.join(save_folder_path, f"{base_filename}error.txt"), "w"
            ) as f:
                error_message = response_body["error"]
                f.write(error_message)

        # Write the images to disk. Note, images are base64 strings representing PNGs.
        if save_folder_path:
            for index, image_base64 in enumerate(response_body.get("images", [])):
                image = base64_to_pil_image(image_base64)
                image.save(
                    os.path.join(save_folder_path, f"{base_filename}image_{index}.png")
                )

        return response_body

    except Exception as ex:
        # Write the error message to disk.
        if save_folder_path:
            with open(
                os.path.join(save_folder_path, f"{base_filename}error.txt"), "w"
            ) as f:
                response_str = str(ex.response)
                f.write(response_str)
        raise ex
